### PR TITLE
Refactor inventory adjustment

### PR DESCRIPTION
**Pull request by mbayopanda**
This PR add a new module for inventory adjustment, it works by resetting all lots of defined inventory and just consider the quantity and the lots which are defined as stock for the inventory selected.

image

This inventory adjustment produce two type of movement flux : 
- Remise à zéro des inventaires : which contains movements of setting to zero inventory lots
- Nouveau stock après inventaire : which contains movements of new stock defined by the user

image

At the end in the lot registry we will have for the selected inventories during the adjustment process, only the defined lot and quantity for the selected inventories.

If an inventory was not selected during the adjustment process its lots will not be affected by the reset.

image

### CODE REVIEW

File: server/models/migrations/next/migrate.sql
Comment (jniles) [inline server/models/migrations/next/migrate.sql]: You forgot a comma here.

**Comment by jniles** [comment]
@mbayopanda let me know when you want a review of this.  It looks like you are still fighting tests ⚔️

**Comment by mbayopanda** [comment]
@jniles you can review it, Travis sometime likes to annoy people

**Comment by jniles** [comment]
@mbayopanda will do once I'm done with my issues.

**Comment by kwilu** [comment]
Heads up @mbayopanda, we just merged some commits that conflict with the changes your made in this pull request! You can review this repository's [recent commits](https://github.com/IMA-WorldHealth/bhima/commits/master) to see where the conflicts occur. Please rebase your feature branch against the `upstream/master` branch and [resolve](https://zulip.readthedocs.io/en/latest/git/troubleshooting.html#recover-from-a-git-rebase-failure) your pull request's merge conflicts accordingly.

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.html
Comment (jniles) [inline client/src/modules/stock/inventory-adjustment/inventory-adjustment.html:19]: I had forgotten we had the `<bh-dropdown-menu>` component.  Cool!

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.js
Comment (jniles) [inline client/src/modules/stock/inventory-adjustment/inventory-adjustment.js]: It looks like this is used in multiple modules.  I think we should probably put this code in a method in the [`StockFormService`](https://github.com/IMA-WorldHealth/bhima/blob/master/client/src/modules/stock/StockForm.service.js#L1) so that we can call `vm.Stock.isValid()` or something.

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.js
Comment (jniles) [inline client/src/modules/stock/inventory-adjustment/inventory-adjustment.js]: Is this just a number?  If so, then you don't need `angular.copy()` - it isn't an object that needs to be de-referenced.

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.html
Comment (jniles) [inline client/src/modules/stock/inventory-adjustment/inventory-adjustment.html:74]: Could you put a loading indicator on here?  When we change the depot, this needs to reload and it takes a little while.

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.html
Comment (jniles) [inline client/src/modules/stock/inventory-adjustment/inventory-adjustment.html]: If we load all stock at the beginning of the operation, do we still need this input group?

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.js
Comment (jniles) [inline client/src/modules/stock/inventory-adjustment/inventory-adjustment.js]: What are selectedLots in the case?  Aren't they all lots?

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.js
Comment (jniles) [inline client/src/modules/stock/inventory-adjustment/inventory-adjustment.js:234]: We need a check here to see if `movement.lots.length === 0` and block submission if that is the case.  See an example below:
x1pOg8y8b7

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.js
Comment (jniles) [inline client/src/modules/stock/inventory-adjustment/inventory-adjustment.js:95]: I don't think we should be filtering on these columns:
image

**Review by jniles — COMMENTED** [review]
@mbayopanda this is the functionality I wanted to see!  Thanks so much!

There is obviously quite a bit of code you can remove since the behavior has evolved a lot from your first implementation.  I've made a few comments, but let me know when you feel ready for me to review the code.

The behavior is nice 👍

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.js
Comment (mbayopanda) [reply client/src/modules/stock/inventory-adjustment/inventory-adjustment.js]: We don't need that now

File: client/src/modules/stock/StockForm.service.js
Comment (jniles) [inline client/src/modules/stock/StockForm.service.js:195]: :+1:

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.html
Comment (jniles) [inline client/src/modules/stock/inventory-adjustment/inventory-adjustment.html:48]: Nice 👍

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.html
Comment (jniles) [inline client/src/modules/stock/inventory-adjustment/inventory-adjustment.html]: It doesn't look like this `suspend()` works or is even defined.

**Comment by mbayopanda** [comment]
@jniles could you review this PR please

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.js
Comment (jniles) [inline client/src/modules/stock/inventory-adjustment/inventory-adjustment.js:134]: Do we need the `addItems()` function still?

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.js
Comment (jniles) [inline client/src/modules/stock/inventory-adjustment/inventory-adjustment.js]: Do we need a `removeItem()` function?

File: client/src/modules/stock/stock.routes.js
Comment (jniles) [inline client/src/modules/stock/stock.routes.js]: I don't think you need the `params` object here.

File: package.json
Comment (jniles) [inline package.json:21]: 👍

File: server/controllers/stock/index.js
Comment (jniles) [inline server/controllers/stock/index.js:198]: 👍

File: server/controllers/stock/reports/stock/adjustment_receipt.js
Comment (jniles) [inline server/controllers/stock/reports/stock/adjustment_receipt.js:36]: Just for clarify, can you make this into `FLUX_TYPE.join(',')`?  I know that be default, it will call `Array.prototype.toString()` and that the `toString()` method joins the array with a comma, but someone reading the code in the future may not know that.

**Review by jniles — COMMENTED** [review]
Just a few more comments!

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.js
Comment (mbayopanda) [reply client/src/modules/stock/inventory-adjustment/inventory-adjustment.js:155]: nope

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.js
Comment (mbayopanda) [reply client/src/modules/stock/inventory-adjustment/inventory-adjustment.js]: nope

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.js
Comment (mbayopanda) [reply client/src/modules/stock/inventory-adjustment/inventory-adjustment.js:134]: It is not used in the template but it used to set lines when we load lots

**Comment by jniles** [comment]
@mbayopanda let me know when you need another review.

**Comment by mbayopanda** [comment]
@jniles yes, i need another review

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.js
Comment (jniles) [inline client/src/modules/stock/inventory-adjustment/inventory-adjustment.js]: I don't think you need the enterprise here.

File: client/src/modules/stock/inventory-adjustment/inventory-adjustment.js
Comment (jniles) [reply client/src/modules/stock/inventory-adjustment/inventory-adjustment.js]: .. can we remove it then?

File: client/src/modules/stock/inventory-adjustment/templates/quantity.tmpl.html
Comment (jniles) [inline client/src/modules/stock/inventory-adjustment/templates/quantity.tmpl.html]: Do you still need this `disabled` check?

File: server/controllers/stock/index.js
Comment (jniles) [inline server/controllers/stock/index.js:222]: You could probably move this check above the `Fiscal.lookup()` call.

File: server/controllers/stock/index.js
Comment (jniles) [inline server/controllers/stock/index.js]: You do this check on the client, right?  Can we just skip this?

File: server/controllers/stock/index.js
Comment (jniles) [inline server/controllers/stock/index.js]: I can't figure out where we will get negative quantities from.  Where do they come from?

File: server/controllers/stock/index.js
Comment (jniles) [inline server/controllers/stock/index.js]: Why do you need an extra check here to get lots?

File: server/controllers/stock/index.js
Comment (jniles) [inline server/controllers/stock/index.js:290]: Nice 👍

File: server/controllers/stock/index.js
Comment (jniles) [inline server/controllers/stock/index.js]: I think you already did the ` (lot.quantity !== lot.oldQuantity)` check once, right?

File: test/end-to-end/stock/stock.adjustment.page.js
Comment (jniles) [inline test/end-to-end/stock/stock.adjustment.page.js:78]: 👍

**Review by jniles — COMMENTED** [review]
Client code looks pretty good now!

Out of curiosity, why did you decide to introduce two new movement types?  Can we just use the regular adjustment types?

File: server/controllers/stock/index.js
Comment (mbayopanda) [reply server/controllers/stock/index.js]: We had found some stock with negative quantities at IMCK in the stock because of some early stock module which allowed to escape validation of stock quantity during the stock exit process

File: server/controllers/stock/index.js
Comment (mbayopanda) [reply server/controllers/stock/index.js]: Yeah i did that on the client, but i was also afraid if somehow these data were not filtered on the client

File: server/controllers/stock/index.js
Comment (jniles) [reply server/controllers/stock/index.js]: Aha!  That makes much more sense.  Can you leave a comment saying that?  I was combing through your code trying to figure out were this would have been possible.

**Comment by mbayopanda** [comment]
@jniles have you time for a review

File: server/models/migrations/next/migrate.sql
Comment (jniles) [inline server/models/migrations/next/migrate.sql]: I think this made it into the previous migration.  Can you remove them please?

**Review by jniles — COMMENTED** [review]
@mbayopanda, just one small comment on the migration script, after that, I think this can be merged.  Any further bugs/refinements can be made in separate PRs.  Well done!

**Review by jniles — APPROVED** [review]
LGTM.

bors r+

**Pull request merged**
